File: python/zpy/tests/plan_oracle.py
"""Plan-driven test oracle — port of ``PlanOracle.scala``.

A ``Plan`` defines N "true" entities, each with K noisy row variants carrying a
CVE-YYYY-NNNNN id in the description (sometimes the summary). The model never
sees ``entity_id``; it is ground truth and a perfect oracle labeller.
"""

from __future__ import annotations

import itertools
import random
from dataclasses import dataclass

import pandas as pd

from zingg_py import features
from zingg_py.labeller import OracleLabeller
from zingg_py.schema import ZinggConf

_PRODUCTS = ["nginx", "openssl", "log4j", "spring-boot", "kubernetes",
             "postgres", "redis", "kafka", "envoy", "grpc"]
_SEVERITIES = ["low", "medium", "high", "critical"]
_PRIORITIES = ["P1", "P2", "P3", "P4"]
_CORE_TEMPLATES = [
    "remote code execution in %s", "buffer overflow affecting %s",
    "authentication bypass for %s", "privilege escalation in %s",
    "denial of service in %s", "memory corruption in %s",
    "SQL injection via %s", "path traversal in %s",
]
_SUMMARY_PREFIXES = ["[SECURITY] ", "URGENT: ", "Vuln: ", "", "Fix needed - ", "Audit finding - "]
_DESC_PREFIXES = ["Reported by scanner. ", "Detected via dependabot. ", "Found in audit. ",
                  "Customer reported: ", "Internal scan: ", ""]


@dataclass(frozen=True)
class PlanEntity:
    entity_id: int
    cve_id: str
    product: str
    core_text: str
    severity: str


@dataclass(frozen=True)
class PlanRow:
    row_id: int
    entity_id: int
    summary: str
    description: str
    priority: str


@dataclass(frozen=True)
class Plan:
    entities: tuple
    rows: tuple

    @property
    def truth(self) -> dict:
        return {r.row_id: r.entity_id for r in self.rows}

    def to_df(self, cfg: ZinggConf) -> pd.DataFrame:
        return pd.DataFrame(
            [{cfg.id_col: r.row_id, "summary": r.summary,
              "description": r.description, "priority": r.priority} for r in self.rows],
            columns=[cfg.id_col, "summary", "description", "priority"],
        )

    @property
    def labeller(self) -> OracleLabeller:
        return OracleLabeller(self.truth)


def _perturb(text: str, rng: random.Random) -> str:
    words = text.split(" ")
    if len(words) > 2 and rng.random() < 0.3:
        i = rng.randrange(len(words) - 1)
        words[i], words[i + 1] = words[i + 1], words[i]
    joined = " ".join(words)
    if rng.random() < 0.3 and len(joined) > 4:
        i = rng.randrange(len(joined))
        return joined[:i] + joined[i + 1:]
    return joined


def build(seed: int, n_entities: int, variants_per_entity: int) -> Plan:
    rng = random.Random(seed)
    entities = []
    for i in range(n_entities):
        year = 2018 + rng.randrange(8)
        num = rng.randrange(90000) + 10000
        cve = f"CVE-{year}-{num:05d}"
        product = _PRODUCTS[rng.randrange(len(_PRODUCTS))]
        core = _CORE_TEMPLATES[rng.randrange(len(_CORE_TEMPLATES))] % product
        severity = _SEVERITIES[rng.randrange(len(_SEVERITIES))]
        entities.append(PlanEntity(i, cve, product, core, severity))

    rows = []
    for e in entities:
        for v in range(variants_per_entity):
            row_id = e.entity_id * 1000 + v
            s_prefix = _SUMMARY_PREFIXES[rng.randrange(len(_SUMMARY_PREFIXES))]
            d_prefix = _DESC_PREFIXES[rng.randrange(len(_DESC_PREFIXES))]
            cve_in_summary = rng.random() < 0.5
            if cve_in_summary:
                summary = f"{s_prefix}{e.cve_id} — {_perturb(e.core_text, rng)}"
            else:
                summary = f"{s_prefix}{_perturb(e.core_text, rng)}"
            description = (f"{d_prefix}{_perturb(e.core_text, rng)}. Affects {e.product}. "
                          f"Severity: {e.severity}. Reference: {e.cve_id}.")
            base_pri = _PRIORITIES[e.entity_id % len(_PRIORITIES)]
            priority = _PRIORITIES[rng.randrange(len(_PRIORITIES))] if rng.random() < 0.2 else base_pri
            rows.append(PlanRow(row_id, e.entity_id, summary, description, priority))

    return Plan(tuple(entities), tuple(rows))


def _pairs_frame(cfg: ZinggConf, pairs, labels) -> pd.DataFrame:
    recs = []
    for (l, r), lab in zip(pairs, labels):
        recs.append({
            f"{ZinggConf.LEFT_PREFIX}{cfg.id_col}": l.row_id,
            f"{ZinggConf.LEFT_PREFIX}summary": l.summary,
            f"{ZinggConf.LEFT_PREFIX}description": l.description,
            f"{ZinggConf.LEFT_PREFIX}priority": l.priority,
            f"{ZinggConf.RIGHT_PREFIX}{cfg.id_col}": r.row_id,
            f"{ZinggConf.RIGHT_PREFIX}summary": r.summary,
            f"{ZinggConf.RIGHT_PREFIX}description": r.description,
            f"{ZinggConf.RIGHT_PREFIX}priority": r.priority,
            cfg.label_col: lab,
        })
    df = pd.DataFrame(recs)
    return features.add_features(df, cfg)


def labeled_training_set(cfg: ZinggConf, plan: Plan, n_negatives: int = 40) -> pd.DataFrame:
    """Positives: ALL same-entity pairs. Negatives: random cross-entity pairs."""
    by_entity: dict = {}
    for r in plan.rows:
        by_entity.setdefault(r.entity_id, []).append(r)

    positives = []
    for rs in by_entity.values():
        positives.extend(itertools.combinations(rs, 2))

    rng = random.Random(0xBEEF)
    negatives = []
    ents = list(plan.entities)
    if len(ents) >= 2:
        for _ in range(n_negatives):
            ea = ents[rng.randrange(len(ents))]
            eb = ents[rng.randrange(len(ents))]
            if ea.entity_id == eb.entity_id:
                continue
            a = by_entity[ea.entity_id][rng.randrange(len(by_entity[ea.entity_id]))]
            b = by_entity[eb.entity_id][rng.randrange(len(by_entity[eb.entity_id]))]
            negatives.append((a, b))

    pairs = list(positives) + negatives
    labels = [1.0] * len(positives) + [0.0] * len(negatives)
    return _pairs_frame(cfg, pairs, labels)


def labeled_training_set_with_noise(cfg: ZinggConf, plan: Plan, n_negatives: int,
                                    noise_fraction: float, noise_seed: int) -> pd.DataFrame:
    clean = labeled_training_set(cfg, plan, n_negatives)
    l_id = f"{ZinggConf.LEFT_PREFIX}{cfg.id_col}"
    r_id = f"{ZinggConf.RIGHT_PREFIX}{cfg.id_col}"

    def flip(row):
        lid, rid, label = int(row[l_id]), int(row[r_id]), row[cfg.label_col]
        r = random.Random(noise_seed ^ (lid * 0x9E3779B97F4A7C15) ^ rid)
        return 1.0 - label if r.random() < noise_fraction else label

    out = clean.copy()
    out[cfg.label_col] = clean.apply(flip, axis=1)
    return out


def pairwise_f1(predicted: dict, truth: dict):
    """Pairwise (precision, recall, f1) between predicted clusters and truth."""
    ids = list(truth.keys())
    tp = fp = fn = 0
    for a, b in itertools.combinations(ids, 2):
        same_truth = truth[a] == truth[b]
        pa, pb = predicted.get(a), predicted.get(b)
        same_pred = pa is not None and pb is not None and pa == pb
        if same_truth and same_pred:
            tp += 1
        elif same_pred and not same_truth:
            fp += 1
        elif same_truth and not same_pred:
            fn += 1
    precision = 1.0 if tp + fp == 0 else tp / (tp + fp)
    recall = 1.0 if tp + fn == 0 else tp / (tp + fn)
    f1 = 0.0 if precision + recall == 0 else 2 * precision * recall / (precision + recall)
    return precision, recall, f1

File: zscala/src/test/scala/zingg/zs/PlanOracle.scala
package zingg.zs

import org.apache.spark.sql.{DataFrame, Row, SparkSession}
import org.apache.spark.sql.types._
import org.scalacheck.Gen

/** Plan-driven test oracle.
  *
  * A `Plan` defines N "true" entities, each with K noisy row variants. Every
  * variant carries an `entity_id` field — this is ground truth and the only
  * thing the test compares against. The model never sees `entity_id`.
  *
  * The plan also serves as a perfect oracle labeller: for any pair of rows,
  * we know whether they should match (same entity_id) or not.
  *
  * The CVE-JIRA scenario: each entity has a CVE-YYYY-NNNNN identifier embedded
  * somewhere in its description (sometimes also in summary). Noisy variants
  * paraphrase, reorder, add prefixes, change priority. The blocking-tree
  * learner should discover the CVE regex hash from labels and use it as a
  * near-perfect block key.
  */
final case class PlanEntity(
    entityId: Int,
    cveId: String,
    product: String,
    coreText: String,
    severity: String
)

final case class PlanRow(
    rowId: Long,
    entityId: Int,
    summary: String,
    description: String,
    priority: String
)

final case class Plan(entities: Seq[PlanEntity], rows: Seq[PlanRow]) {

  /** True cluster assignment: rowId -> entityId. */
  def truth: Map[Long, Int] = rows.iterator.map(r => r.rowId -> r.entityId).toMap

  /** Convert to a Spark DataFrame WITHOUT the entity_id (the model never sees it). */
  def toDF(spark: SparkSession, cfg: ZinggConf): DataFrame = {
    val schema = StructType(Seq(
      StructField(cfg.idCol,     LongType,   nullable = false),
      StructField("summary",     StringType, nullable = true),
      StructField("description", StringType, nullable = true),
      StructField("priority",    StringType, nullable = true)
    ))
    val rdd = spark.sparkContext.parallelize(
      rows.map(r => Row(r.rowId, r.summary, r.description, r.priority)),
      numSlices = 2
    )
    spark.createDataFrame(rdd, schema)
  }

  /** Build an `OracleLabeller` wired to this plan's ground truth. */
  def labeller: OracleLabeller = new OracleLabeller(truth)
}

object Plan {

  private val products = Vector("nginx", "openssl", "log4j", "spring-boot", "kubernetes",
                                 "postgres", "redis", "kafka", "envoy", "grpc")
  private val severities = Vector("low", "medium", "high", "critical")
  private val priorities = Vector("P1", "P2", "P3", "P4")

  private val coreTemplates = Vector(
    "remote code execution in %s",
    "buffer overflow affecting %s",
    "authentication bypass for %s",
    "privilege escalation in %s",
    "denial of service in %s",
    "memory corruption in %s",
    "SQL injection via %s",
    "path traversal in %s"
  )

  private val summaryPrefixes = Vector(
    "[SECURITY] ", "URGENT: ", "Vuln: ", "", "Fix needed - ", "Audit finding - "
  )

  private val descPrefixes = Vector(
    "Reported by scanner. ", "Detected via dependabot. ", "Found in audit. ",
    "Customer reported: ", "Internal scan: ", ""
  )

  /** Generate one random plan with `nEntities` entities and `variantsPerEntity` rows each. */
  def gen(nEntities: Int, variantsPerEntity: Int): Gen[Plan] = for {
    seed <- Gen.choose(1L, Long.MaxValue)
  } yield build(seed, nEntities, variantsPerEntity)

  /** Deterministic build given a seed. */
  def build(seed: Long, nEntities: Int, variantsPerEntity: Int): Plan = {
    val rng = new scala.util.Random(seed)
    val entities = (0 until nEntities).map { i =>
      val year = 2018 + rng.nextInt(8)
      val num  = rng.nextInt(90000) + 10000
      val cve  = f"CVE-$year-$num%05d"
      val product  = products(rng.nextInt(products.size))
      val core     = coreTemplates(rng.nextInt(coreTemplates.size)).format(product)
      val severity = severities(rng.nextInt(severities.size))
      PlanEntity(i, cve, product, core, severity)
    }

    val rows = entities.flatMap { e =>
      (0 until variantsPerEntity).map { v =>
        val rowId = (e.entityId.toLong * 1000L) + v.toLong
        val sPrefix = summaryPrefixes(rng.nextInt(summaryPrefixes.size))
        val dPrefix = descPrefixes(rng.nextInt(descPrefixes.size))
        // Vary where the CVE id appears: always in description, sometimes in summary.
        val cveInSummary = rng.nextBoolean()
        val summary =
          if (cveInSummary) s"$sPrefix${e.cveId} — ${perturb(e.coreText, rng)}"
          else              s"$sPrefix${perturb(e.coreText, rng)}"
        val description =
          s"$dPrefix${perturb(e.coreText, rng)}. Affects ${e.product}. " +
          s"Severity: ${e.severity}. Reference: ${e.cveId}."
        // Priority is mostly stable per entity, sometimes drifts.
        val basePri = priorities(e.entityId % priorities.size)
        val priority =
          if (rng.nextDouble() < 0.2) priorities(rng.nextInt(priorities.size))
          else basePri
        PlanRow(rowId, e.entityId, summary, description, priority)
      }
    }

    Plan(entities, rows)
  }

  private def perturb(text: String, rng: scala.util.Random): String = {
    val words = text.split(" ").toBuffer
    // Occasional word swap.
    if (words.size > 2 && rng.nextDouble() < 0.3) {
      val i = rng.nextInt(words.size - 1)
      val tmp = words(i); words(i) = words(i + 1); words(i + 1) = tmp
    }
    // Occasional single-char typo.
    val joined = words.mkString(" ")
    if (rng.nextDouble() < 0.3 && joined.length > 4) {
      val i = rng.nextInt(joined.length)
      joined.substring(0, i) + joined.substring(i + 1)
    } else joined
  }

  /** Build a labeled training DataFrame directly from a plan, in the shape
    * `Classifier.train` expects: left-prefixed / right-prefixed columns + z_features + z_label.
    *
    * Positives: ALL same-entity pairs.
    * Negatives: `nNegatives` random cross-entity pairs.
    */
  def labeledTrainingSet(spark: SparkSession, cfg: ZinggConf,
                         plan: Plan, nNegatives: Int = 40): DataFrame = {
    import org.apache.spark.sql.functions.{lit, udf, struct, col}
    import org.apache.spark.ml.linalg.Vectors

    val byEntity = plan.rows.groupBy(_.entityId)
    val positives: Seq[(PlanRow, PlanRow)] = byEntity.values.flatMap { rs =>
      val v = rs.toVector
      for { i <- v.indices; j <- (i + 1) until v.size } yield (v(i), v(j))
    }.toSeq

    val rng = new scala.util.Random(0xBEEF)
    val negatives: Seq[(PlanRow, PlanRow)] =
      if (plan.entities.size < 2) Seq.empty
      else (0 until nNegatives).flatMap { _ =>
        val eA = plan.entities(rng.nextInt(plan.entities.size))
        val eB = plan.entities(rng.nextInt(plan.entities.size))
        if (eA.entityId == eB.entityId) None
        else {
          val a = byEntity(eA.entityId)(rng.nextInt(byEntity(eA.entityId).size))
          val b = byEntity(eB.entityId)(rng.nextInt(byEntity(eB.entityId).size))
          Some((a, b))
        }
      }

    val schema = StructType(Seq(
      StructField(s"${ZinggConf.LeftPrefix}${cfg.idCol}",        LongType),
      StructField(s"${ZinggConf.LeftPrefix}summary",              StringType),
      StructField(s"${ZinggConf.LeftPrefix}description",          StringType),
      StructField(s"${ZinggConf.LeftPrefix}priority",             StringType),
      StructField(s"${ZinggConf.RightPrefix}${cfg.idCol}",       LongType),
      StructField(s"${ZinggConf.RightPrefix}summary",             StringType),
      StructField(s"${ZinggConf.RightPrefix}description",         StringType),
      StructField(s"${ZinggConf.RightPrefix}priority",            StringType),
      StructField(cfg.labelCol,                                   DoubleType)
    ))
    def mkRow(l: PlanRow, r: PlanRow, label: Double): Row =
      Row(l.rowId, l.summary, l.description, l.priority,
          r.rowId, r.summary, r.description, r.priority, label)

    val rows = positives.map { case (l, r) => mkRow(l, r, 1.0) } ++
               negatives.map { case (l, r) => mkRow(l, r, 0.0) }
    val rdd = spark.sparkContext.parallelize(rows, 2)
    val pairs = spark.createDataFrame(rdd, schema)
    Features.addFeatures(pairs, cfg)
  }

  /** Same as `labeledTrainingSet` but flips each label independently with
    * probability `noiseFraction`. Simulates a fallible labeller (human or LLM).
    */
  def labeledTrainingSetWithNoise(spark: SparkSession, cfg: ZinggConf, plan: Plan,
                                  nNegatives: Int, noiseFraction: Double,
                                  noiseSeed: Long): DataFrame = {
    import org.apache.spark.sql.functions.{col, lit, udf, when}
    val clean = labeledTrainingSet(spark, cfg, plan, nNegatives)
    val rng = new scala.util.Random(noiseSeed)
    // Per-row deterministic flip: hash row id + seed → uniform [0,1).
    val nseed = noiseSeed
    val lFrac = noiseFraction
    val flip = udf { (lId: Long, rId: Long, label: Double) =>
      val r = new scala.util.Random(nseed ^ (lId * 0x9E3779B97F4A7C15L) ^ rId)
      if (r.nextDouble() < lFrac) 1.0 - label else label
    }
    val lId = s"${ZinggConf.LeftPrefix}${cfg.idCol}"
    val rId = s"${ZinggConf.RightPrefix}${cfg.idCol}"
    clean.withColumn(cfg.labelCol, flip(col(lId), col(rId), col(cfg.labelCol)))
  }
}

/** Evaluation: pairwise precision/recall vs ground truth. */
object Evaluator {

  /** Compute pairwise F1 between predicted clusters and ground truth.
    *
    *  Pairwise metrics: over all unordered pairs of rows,
    *    TP = pairs predicted-same AND truly-same
    *    FP = pairs predicted-same AND truly-different
    *    FN = pairs predicted-different AND truly-same
    */
  def pairwiseF1(predicted: Map[Long, Any], truth: Map[Long, Int]): (Double, Double, Double) = {
    val ids = truth.keys.toVector
    var tp, fp, fn = 0L
    for {
      i <- ids.indices
      j <- (i + 1) until ids.size
    } {
      val a = ids(i); val b = ids(j)
      val sameTruth = truth(a) == truth(b)
      val samePred  = predicted.get(a).exists(p => predicted.get(b).contains(p))
      (sameTruth, samePred) match {
        case (true,  true)  => tp += 1
        case (false, true)  => fp += 1
        case (true,  false) => fn += 1
        case _              => ()
      }
    }
    val precision = if (tp + fp == 0) 1.0 else tp.toDouble / (tp + fp)
    val recall    = if (tp + fn == 0) 1.0 else tp.toDouble / (tp + fn)
    val f1        = if (precision + recall == 0.0) 0.0 else 2 * precision * recall / (precision + recall)
    (precision, recall, f1)
  }
}
